@xen-orchestra/lite/src/components/Modal.tsx

```diff
@@ -0,0 +1,116 @@
+import React from 'react'
+import { FormattedMessage } from 'react-intl'
+import { withState } from 'reaclette'
+
+import Button from './Button'
+
+interface GeneralPropsModal {
+  message: JSX.Element
+  title: JSX.Element
+}
+
+interface PropsModal extends GeneralPropsModal {
+  onReject: (reason: unknown) => void
+  onSuccess: (value: unknown) => void
+}
+
+export const confirm = ({ message, title }: GeneralPropsModal) =>
+  new Promise((resolve, reject) => modal({ message, onReject: reject, onSuccess: resolve, title }))
+
+const modal = ({ message, onReject, onSuccess, title }: PropsModal) => {
+  if (instance === undefined) {
+    throw new Error('No modal instance')
+  }
+
+  instance.state.message = message
+  instance.state.onReject = onReject
+  instance.state.onSuccess = onSuccess
+  instance.state.showModal = true
+  instance.state.title = title
+}
+
+const STYLES_MODAL: React.CSSProperties = {
+  backgroundColor: 'lightgrey',
+  border: '1px solid black',
+  left: '50%',
+  position: 'absolute',
+  top: '20%',
+  transform: 'translate(-50%,-50%)',
+}
+
+interface ParentState {}
+
+interface State {
+  message?: JSX.Element
+  onReject?: (reason: unknown) => void
+  onSuccess?: (value: unknown) => void
+  showModal: boolean
+  title?: JSX.Element
+}
+
+interface Props {}
+
+interface ParentEffects {}
+
+interface Effects {
+  _closeModal: () => void
+  _reject: () => void
+  _success: () => void
+}
+
+interface Computed {}
+
+let instance: EffectContext<State, Props, Effects, Computed, ParentState, ParentEffects> | undefined
+const Modal = withState<State, Props, Effects, Computed, ParentState, ParentEffects>(
+  {
+    initialState: () => ({
+      message: undefined,
+      onReject: undefined,
+      onSuccess: undefined,
+      showModal: false,
+      title: undefined,
+    }),
+    effects: {
+      initialize: function () {
+        if (instance !== undefined) {
+          throw new Error('Modal is a singelton')
+        }
+        instance = this
+      },
+      finalize: function () {
+        instance = undefined
+      },
+      _closeModal: function () {
+        this.state.showModal = false
+      },
+      _reject: function () {
+        this.state.onReject!('Normal reject')
```

Comment:
Why `!`?

Reply:
We declare in the [State interface](https://github.com/vatesfr/xen-orchestra/pull/5722/files#diff-b9ed30d768e04236b2eb058c12f1925ebb2b966f8c6d328b36db5e434ae7227dR45) `onReject` can be potentially undefined because when the Modal is mounted, we don't know what value to [assign to it.](https://github.com/vatesfr/xen-orchestra/pull/5722/files#diff-b9ed30d768e04236b2eb058c12f1925ebb2b966f8c6d328b36db5e434ae7227dR68)

To avoid `!` we can do something like that:
```typescript
interface State {
  onReject: (reason: unknown) => void
  // ...
}

initialState: () => ({
 onReject: () => {}
})

```

Reply:
Why don't you simply not call it when it's `undefined`? You can even do this with an optional chaining: `this.state.onReject?.()`. Having an `undefined` `onReject` shouldn't be a problem IMO. Also, when you're using `!`, you're telling TypeScript that `onReject` isn't `undefined`. But you actually don't know that, so you're potentially trying to call `undefined` there.